### ISSUE TITLE
Paste vs bracketed paste vs source; echo vs silent.

### ISSUE THREAD

**Issue by jalvesaq** [jalvesaq/tmp-R-Nvim#23]
In the past, when sending multiple lines to RConsole, if the total number of bytes was over 4 Kb, the text was incompletely pasted by Tmux (required to run R in an external terminal on Unix). To test if the problem persists, I converted a PDF book to txt and then converted each line into a  `cat()`:

```
%s/\(.*\)/cat("\1\\n")/
```

The result was a file with 500 Kb.

There was no problem sending all the lines concatenated with `"\n"` directly to R if it was running in Neovim's built-in terminal. Also, there was no problem sending the code to Tmux.

However, a bracketed paste didn't work with Neovim's built-in terminal or Tmux.

Considering these results we may consider that we no longer need to source multiple lines if running R on a local Linux system. However, it is prudent to keep the option to source code at least in some circumstances, for example, when running R in a remote machine. We also have to keep the bracketed paste option for Python code in Quarto documents.

Although it's possible, I can't see any usefulness of pasting hundreds of lines on the R Console since the code will scroll up and be out of sight anyway. So, I'm trying to figure out what should be the default behavior. Perhaps: Paste the code directly if the number of lines is below `max_lines_to_paste`; otherwise, source it. 

Another problem is the number of key bindings created by default: four for each source operation, combining "echo" and "movement". I think we can get rid at least of the "echo" option because `echo=TRUE` can be added to `source_args`.

@PMassicotte , @she3o, any suggestions?

**Comment by PMassicotte** [comment]
I like the idea of `max_line_to_paste`. We could also echo like the first n and the last n lines of the block. So, if I understand correctly, we could _simply_ send the lines to the console without sourcing a temp file (except for some use case you describe)?

**Comment by jalvesaq** [comment]
> We could also echo like the first n and the last n lines of the block.

It's a good idea, but I don't know how to implement it with `base::source()`.

> So, if I understand correctly, we could _simply_ send the lines to the console without sourcing a temp file (except for some use case you describe)?

Yes, except if the number of lines is `> max_lines_paste`. I suggest a default value close to 20 for this option. Of course, users would be free to change it to any value between 1 and `inf`.

**Comment by jalvesaq** [comment]
> We could also echo like the first n and the last n lines of the block.

Actually, this could be done inside `nvimcom:::NvimR.source()`

**Comment by PMassicotte** [comment]
Should send functions work at the moment? I am trying to use the tmp-R-Nvim and I keep getting *R is not ready yet*, coming from this line

https://github.com/jalvesaq/tmp-R-Nvim/blob/587be7749934431b5ab13e2745eeaa4f14560441/lua/r/send.lua#L65

**Comment by jalvesaq** [comment]
I was testing the external terminal, and introduced this bug for the built-in terminal.  I will fix it in a couple of minutes.

**Comment by PMassicotte** [comment]
No rush :)

**Comment by jalvesaq** [comment]
I can't replicate the bug.

**Comment by jalvesaq** [comment]
Could you try this:

```sh
echo 'remove.packages("nvimcom")' | R
```
Maybe nvimcom needs to be rebuilt...

**Comment by PMassicotte** [comment]
I removed nvimcom and it re-install when I open a R file in nvimr. But I still have that same issue.

![Peek 2024-02-10 13-08](https://github.com/jalvesaq/tmp-R-Nvim/assets/4519221/24505bd8-3812-4187-810e-c0325334bacb)

**Comment by PMassicotte** [comment]
WIll investigate and report back.

**Comment by PMassicotte** [comment]
Actually, I am getting some lua errors. Do you get some if you type `:messages`

![image](https://github.com/jalvesaq/tmp-R-Nvim/assets/4519221/e8ad087d-3e38-4ff0-9ef3-3e3e0a3ffdcf)

**Comment by PMassicotte** [comment]
Looks like it is coming from `exec_stdout_cmd`

**Comment by jalvesaq** [comment]
When an environment variable doesn't exist, the `vim.env.VAR` is `nil`...
I will grep `'vim.env.*\.\.'` and fix the cases when the environment variable might not exist.

**Comment by jalvesaq** [comment]
If indeed the problem was the environment variable being `nil`, it should be fixed now.

**Comment by PMassicotte** [comment]
Super, working perfectly fine!

**Comment by PMassicotte** [comment]
The `echo/silent` and `down/stay` are inverted in function call:

https://github.com/jalvesaq/tmp-R-Nvim/blob/46ef896d83331879ec2685628688f2a04f114a50/lua/r/maps.lua#L150-L154

https://github.com/jalvesaq/tmp-R-Nvim/blob/46ef896d83331879ec2685628688f2a04f114a50/lua/r/send.lua#L172-L192

so `m` is equall to either `echo/silent`. 

We can fix that later when we decide what to do with the `echo/silent` options.

**Comment by jalvesaq** [comment]
> We can fix that later when we decide what to do with the echo/silent options.

What is your opinion on this? I prefer to eliminate the "echo" parameter from the send functions because users can set the value of `source_args` as `", echo=TRUE"`. It's redundant to have both maps for echoing what is being sourced and the option to do this with `source_args`. I prefer to simplify the functions and reduce the number of key bindings. But if you like the parameter we can keep it.

**Comment by jalvesaq** [comment]
> Super, working perfectly fine!

If you do `git pull` now the Object Browser will auto start faster because I removed the 1 second delay on Unix.

**Comment by PMassicotte** [comment]
> > We can fix that later when we decide what to do with the echo/silent options.
> 
> What is your opinion on this? I prefer to eliminate the "echo" parameter from the send functions because users can set the value of `source_args` as `", echo=TRUE"`. It's redundant to have both maps for echoing what is being sourced and the option to do this with `source_args`. I prefer to simplify the functions and reduce the number of key bindings. But if you like the parameter we can keep it.

Make total sens. I up to remove echo.

**Comment by PMassicotte** [comment]
> > Super, working perfectly fine!
> 
> If you do `git pull` now the Object Browser will auto start faster because I removed the 1 second delay on Unix.

Oh this is nice. Open a nanosecond after the terminal. Tx!

**Comment by jalvesaq** [comment]
So, please, feel free to remove the "echo" parameter if you have time.

**Comment by PMassicotte** [comment]
all right, will do it right now

**Comment by jalvesaq** [comment]
I uploaded some bug fixes in many files. So, please, do `git pull` before pushing new changes. I intended to do a pull request, but I forgot to create a branch and pushed the changes to the main branch.

**Comment by jalvesaq** [comment]
I'm implementing the option `max_paste_lines`...

**Comment by PMassicotte** [comment]
Great! The project is taking shape!

**Comment by jalvesaq** [comment]
I already tried most features, and they are working. One problem is some key bindings not being activated. When the same function is mapped twice, one with argument `false` and the other with `true`, only the second one is active, as we can see with `:nmap`. This happens at least with SendParagraph, and SendSelection.

**Comment by PMassicotte** [comment]
Meaning that 

```lua
    -- Paragraph
    create_maps("ni", "RSendParagraph",   "pp", "<Cmd>lua require('r.send').paragraph(false)")
    create_maps("ni", "RDSendParagraph",  "pd", "<Cmd>lua require('r.send').paragraph(true)")
```

Is only mapping the last one?

**Comment by jalvesaq** [comment]
Thanks for looking at that!

I found the problem: it was my configuration! I was setting custom key bindings and they were preventing `maps.lua` to do its job.

My default mapping to send lines of code to R, `<Enter>` is also preventing me from using the LSP quickfix list. I will create a new hook to overcome this problem.

**Comment by PMassicotte** [comment]
oh cool!

**Comment by jalvesaq** [comment]
My configuration now is (commented lines ommited):

```lua
return {
    dir = '~/src/tmp-R-Nvim',
    config = {
        R_args = {'--quiet', '--no-save'},
        hook = {
            after_config = function ()
                vim.api.nvim_buf_set_keymap(0, "n", "<Enter>", "<Plug>RDSendLine", {})
                vim.api.nvim_buf_set_keymap(0, "v", "<Enter>", "<Plug>RSendSelection", {})
            end
        },
        -- user_maps_only = true,
        auto_start = 1,
        objbr_auto_start = true,
    },
    lazy = false
}
```

**Comment by PMassicotte** [comment]
Great, I will try that when I am back at home in few hours.

**Comment by jalvesaq** [comment]
I have had to create some maps conditionally:

```lua
    {
        dir = '~/src/tmp-R-Nvim',
        opts = {
            R_args = {'--quiet', '--no-save'},
            hook = {
                after_config = function ()
                    vim.api.nvim_buf_set_keymap(0, 'n', '<LocalLeader>rq', '<Plug>RClose', {})
                    vim.api.nvim_buf_set_keymap(0, 'n', '<LocalLeader>ss', '<Plug>RSendSelection', {})
                    vim.api.nvim_buf_set_keymap(0, 'n', '<LocalLeader>sp', '<Plug>RDSendParagraph', {})
                    vim.api.nvim_buf_set_keymap(0, 'n', '<LocalLeader>rl', '<Plug>RListSpace', {})
                    vim.api.nvim_buf_set_keymap(0, 'n', '<LocalLeader>rm', '<Plug>RClearAll', {})
                    vim.api.nvim_buf_set_keymap(0, 'n', '<LocalLeader>rp', '<Plug>RObjectPr', {})
                    vim.api.nvim_buf_set_keymap(0, 'n', '<LocalLeader>rn', '<Plug>RObjectNames', {})
                    vim.api.nvim_buf_set_keymap(0, 'n', '<LocalLeader>rt', '<Plug>RObjectStr', {})
                    vim.api.nvim_buf_set_keymap(0, 'n', '<LocalLeader>rv', '<Plug>RViewDF', {})
                    vim.api.nvim_buf_set_keymap(0, 'n', '<LocalLeader>ra', '<Plug>RShowArgs', {})
                    vim.api.nvim_buf_set_keymap(0, 'n', '<LocalLeader>rh', '<Plug>RHelp', {})
                    vim.api.nvim_buf_set_keymap(0, 'n', '<LocalLeader>rs', '<Plug>RSummary', {})
                    vim.api.nvim_buf_set_keymap(0, 'n', '<LocalLeader>ro', '<Plug>RUpdateObjBrowser', {})
                    vim.api.nvim_buf_set_keymap(0, 'n', '<LocalLeader>r=', '<Plug>ROpenLists', {})
                    vim.api.nvim_buf_set_keymap(0, 'n', '<LocalLeader>r-', '<Plug>RCloseLists', {})
                    if vim.o.syntax ~= "rbrowser" then
                        vim.api.nvim_buf_set_keymap(0, "n", "<Enter>", "<Plug>RDSendLine", {})
                        vim.api.nvim_buf_set_keymap(0, "v", "<Enter>", "<Plug>RSendSelection", {})
                        vim.api.nvim_buf_set_keymap(0, 'n', '<LocalLeader>sb', '<Plug>RSendMBlock', {})
                        vim.api.nvim_buf_set_keymap(0, 'n', '<LocalLeader>rf', '<Plug>RStart', {})
                    end
                    if vim.o.filetype == "rmd" or vim.o.filetype == "quarto" then
                        vim.api.nvim_buf_set_keymap(0, 'n', '<LocalLeader>kn', '<Plug>RKnit', {})
                        vim.api.nvim_buf_set_keymap(0, 'n', '<LocalLeader>kr', '<Plug>RMakeRmd', {})
                        vim.api.nvim_buf_set_keymap(0, 'n', '<LocalLeader>gn', '<Plug>RNextRChunk', {})
                        vim.api.nvim_buf_set_keymap(0, 'n', '<LocalLeader>gN', '<Plug>RPreviousRChunk', {})
                        vim.api.nvim_buf_set_keymap(0, 'n', '<LocalLeader>sc', '<Plug>RSendChunk', {})
                        vim.api.nvim_buf_set_keymap(0, 'n', '<LocalLeader>sh', '<Plug>RSendChunkFH', {})
                    end
                end
            },
            user_maps_only = true,
            -- clear_line = false,
            auto_start = 1,
            -- bracketed_paste = false,
            objbr_auto_start = true,
            term_title = "tmux",
            -- external_term = "foot",
        },
        lazy = false
    },
```

**Comment by PMassicotte** [comment]
You have to do it to avoid duplicated map? On my side all works fine (maybe I missed something)

**Comment by jalvesaq** [comment]
The main reason was that when I pressed `<leader>gd` over a function name in a Lua buffer, the LSP opened a quickfix window with two lines and, then, when I pressed `<Enter>` nothing happened. The problem was that I mapped `<Enter>` to send lines to R, and the map was active in all buffers. The hooks are very cheap to add; just 2 lines of code: one in the table `config` and the other where the function should be called.
